nexus/types/src/deployment.rs

```diff
@@ -480,6 +481,104 @@ impl Blueprint {
+        // NTP zone, but some test or test systems may have 0 if they have only
+        // a single sled and that sled's boundary NTP zone is being upgraded.)
+        self.all_omicron_zones(BlueprintZoneDisposition::any).find_map(
+            |(_sled_id, zone)| match &zone.zone_type {
+                BlueprintZoneType::BoundaryNtp(ntp_config) => {
+                    Some(UpstreamNtpConfig {
+                        ntp_servers: &ntp_config.ntp_servers,
+                        dns_servers: &ntp_config.dns_servers,
+                        domain: ntp_config.domain.as_deref(),
+                    })
+                }
+                _ => None,
+            },
+        )
+    }
+
+    /// Return the operator-specified configuration of Nexus.
+    ///
+    /// This information should be operator-configurable, but currently is not:
+    /// we carry it forward from rack setup time onward from blueprint to
+    /// blueprint. Fixing this is
+    /// <https://github.com/oxidecomputer/omicron/issues/9040>.
+    ///
+    /// Returns `None` if this blueprint contains no Nexus zones from which we
+    /// can infer the configuration. (This should only be the case for test
+    /// blueprints - real systems always deploy at least one Nexus zone).
+    pub fn operator_nexus_config(&self) -> Option<OperatorNexusConfig<'_>> {
+        // The Nexus config can't be changed, so it's fine to use
+        // `find()` here and include searching both in-service and expunged
+        // zones. (Real racks will always have at least one in-service Nexus
+        // zone - the one calling this code - but some tests create blueprints
+        // without any.)
+        self.all_omicron_zones(BlueprintZoneDisposition::any).find_map(
+            |(_sled_id, zone)| match &zone.zone_type {
+                BlueprintZoneType::Nexus(nexus_config) => {
+                    Some(OperatorNexusConfig {
+                        external_tls: nexus_config.external_tls,
+                        external_dns_servers: &nexus_config
+                            .external_dns_servers,
+                    })
+                }
+                _ => None,
+            },
+        )
+    }
+
+    /// Returns the complete set of external IP addresses assigned to external
+    /// DNS servers described by this blueprint, including both in-service and
+    /// expunged external DNS zones.
+    ///
+    /// This information should be operator-configurable, but currently is not:
+    /// we carry it forward from rack setup time onward from blueprint to
+    /// blueprint. Fixing this is
+    /// <https://github.com/oxidecomputer/omicron/issues/9040>.
+    ///
+    /// Returns an empty set if this blueprint contains no external DNS zones.
+    /// (This should only be the case for test blueprints - real systems always
+    /// deploy at least one external DNS zone).
+    pub fn all_external_dns_external_ips(&self) -> BTreeSet<IpAddr> {
```

Comment:
I'm assuming that:

- It's possible to have multiple zones with the same external IP as long as no more than one of them is _not_ expunged.
- We do need to include expunged zones because they may have had an IP that we haven't put back into service in a new zone yet.

In that case this code looks right and it's important that it's a Set and not a Vec or something.  This might be worth commenting, but not a big deal.